src/nbl/ui/CWindowWin32.cpp

```diff
@@ -369,6 +367,120 @@ namespace ui
+				RAWMOUSE rawMouse = rawInput.data.mouse;
+ 
+				if ((rawMouse.usFlags & MOUSE_MOVE_RELATIVE) == MOUSE_MOVE_RELATIVE && (rawMouse.lLastX != 0 || rawMouse.lLastY != 0))
+				{
+					SMouseEvent event;
+					event.type = SMouseEvent::EMT_MOTION;
+					event.motionEvent.motionX = rawMouse.lLastX;
+					event.motionEvent.motionY = rawMouse.lLastY;
+					event.window = window;
+					auto lk = inputChannel->lockBackgroundBuffer();
+					lk.lock();
+					inputChannel->pushIntoBackground(std::move(event));
+				}
+				if (rawMouse.usButtonFlags & RI_MOUSE_LEFT_BUTTON_DOWN)
+				{
+					SMouseEvent event;
+					event.type = SMouseEvent::EMT_BUTTON;
+					event.buttonEvent.state = SMouseEvent::SButtonEvent::ES_PRESSED;
+					event.buttonEvent.button = ui::E_MOUSEBUTTON::EMB_LEFT_BUTTON;
+					auto lk = inputChannel->lockBackgroundBuffer();
+					lk.lock();
+					inputChannel->pushIntoBackground(std::move(event));
+				}
+				else if (rawMouse.usButtonFlags & RI_MOUSE_LEFT_BUTTON_UP)
+				{
+					SMouseEvent event;
+					event.type = SMouseEvent::EMT_BUTTON;
+					event.buttonEvent.state = SMouseEvent::SButtonEvent::ES_RELEASED;
+					event.buttonEvent.button = ui::E_MOUSEBUTTON::EMB_LEFT_BUTTON;
+					auto lk = inputChannel->lockBackgroundBuffer();
+					lk.lock();
+					inputChannel->pushIntoBackground(std::move(event));
+				}
+				if (rawMouse.usButtonFlags & RI_MOUSE_RIGHT_BUTTON_DOWN)
+				{
+					SMouseEvent event;
+					event.type = SMouseEvent::EMT_BUTTON;
+					event.buttonEvent.state = SMouseEvent::SButtonEvent::ES_PRESSED;
+					event.buttonEvent.button = ui::E_MOUSEBUTTON::EMB_RIGHT_BUTTON;
+					auto lk = inputChannel->lockBackgroundBuffer();
+					lk.lock();
+					inputChannel->pushIntoBackground(std::move(event));
+				}
+				else if (rawMouse.usButtonFlags & RI_MOUSE_RIGHT_BUTTON_UP)
+				{
+					SMouseEvent event;
+					event.type = SMouseEvent::EMT_BUTTON;
+					event.buttonEvent.state = SMouseEvent::SButtonEvent::ES_RELEASED;
+					event.buttonEvent.button = ui::E_MOUSEBUTTON::EMB_RIGHT_BUTTON;
+					auto lk = inputChannel->lockBackgroundBuffer();
+					lk.lock();
+					inputChannel->pushIntoBackground(std::move(event));
+				}
+				if (rawMouse.usButtonFlags & RI_MOUSE_MIDDLE_BUTTON_DOWN)
+				{
+					SMouseEvent event;
+					event.type = SMouseEvent::EMT_BUTTON;
+					event.buttonEvent.state = SMouseEvent::SButtonEvent::ES_PRESSED;
+					event.buttonEvent.button = ui::E_MOUSEBUTTON::EMB_MIDDLE_BUTTON;
+					auto lk = inputChannel->lockBackgroundBuffer();
+					lk.lock();
+					inputChannel->pushIntoBackground(std::move(event));
+				}
+				else if (rawMouse.usButtonFlags & RI_MOUSE_MIDDLE_BUTTON_UP)
+				{
+					SMouseEvent event;
+					event.type = SMouseEvent::EMT_BUTTON;
+					event.buttonEvent.state = SMouseEvent::SButtonEvent::ES_RELEASED;
+					event.buttonEvent.button = ui::E_MOUSEBUTTON::EMB_MIDDLE_BUTTON;
+					auto lk = inputChannel->lockBackgroundBuffer();
+					lk.lock();
```

Comment:
dont call `lock()`, it's already locked. Same for other cases

Reply:
@sadiuk read up on C++11 RAII lock wrappers